hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterFsck.java

```diff
@@ -0,0 +1,157 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdfs.server.federation.router;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.InetAddress;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.hdfs.server.federation.resolver.FederationNamenodeServiceState;
+import org.apache.hadoop.hdfs.server.federation.store.MembershipStore;
+import org.apache.hadoop.hdfs.server.federation.store.StateStoreService;
+import org.apache.hadoop.hdfs.server.federation.store.protocol.GetNamenodeRegistrationsRequest;
+import org.apache.hadoop.hdfs.server.federation.store.protocol.GetNamenodeRegistrationsResponse;
+import org.apache.hadoop.hdfs.server.federation.store.records.MembershipState;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.util.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Wrapper for the Router to offer the Namenode FSCK.
+ */
+@InterfaceAudience.Private
+public class RouterFsck {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(RouterFsck.class.getName());
+
+  private final Router router;
+  private final InetAddress remoteAddress;
+  private final PrintWriter out;
+  private final Map<String, String[]> pmap;
+
+  public RouterFsck(Router router, Map<String, String[]> pmap,
+                    PrintWriter out, InetAddress remoteAddress) {
+    this.router = router;
+    this.remoteAddress = remoteAddress;
+    this.out = out;
+    this.pmap = pmap;
+  }
+
+  public void fsck() {
+    final long startTime = Time.monotonicNow();
+    try {
+      String msg = "Federated FSCK started by " +
+          UserGroupInformation.getCurrentUser() + " from " + remoteAddress +
+          " at " + new Date();
+      LOG.info(msg);
+      out.println(msg);
+
+      // Check each Namenode in the federation
+      StateStoreService stateStore = router.getStateStore();
+      MembershipStore membership =
+          stateStore.getRegisteredRecordStore(MembershipStore.class);
+      GetNamenodeRegistrationsRequest request =
+          GetNamenodeRegistrationsRequest.newInstance();
+      GetNamenodeRegistrationsResponse response =
+          membership.getNamenodeRegistrations(request);
+      List<MembershipState> memberships = response.getNamenodeMemberships();
+      Collections.sort(memberships);
+      for (MembershipState nn : memberships) {
+        if (nn.getState() == FederationNamenodeServiceState.ACTIVE) {
+          try {
+            String webAddress = nn.getWebAddress();
+            out.write("Checking " + nn + " at " + webAddress + "\n");
```

Comment:
Can we check for this with the webAddress in the unit test?

Reply:
Added the check in the unit test